server/src/main/java/org/elasticsearch/plugins/PluginsUtils.java

```diff
@@ -77,43 +79,111 @@ public static List<Path> findPluginDirs(final Path rootPath) throws IOException
+                    );
+                }
+
+                // case: Major version mismatch
+                if (pluginElasticsearchVersion.major != currentElasticsearchVersion.major) {
+                    throw new IllegalArgumentException(
+                        "Stable Plugin ["
+                            + info.getName()
+                            + "] was built for Elasticsearch major version "
+                            + pluginElasticsearchVersion.major
+                            + " but version "
+                            + Build.current().version()
+                            + " is running"
+                    );
+                }
+
+                // case: stable plugin from the future
+                if (pluginElasticsearchVersion.after(currentElasticsearchVersion)) {
+                    throw new IllegalArgumentException(
+                        "Stable Plugin ["
+                            + info.getName()
+                            + "] was built for Elasticsearch version "
+                            + info.getElasticsearchVersion()
+                            + " but earlier version "
+                            + Build.current().version()
+                            + " is running"
+                    );
+                }
+            } else if (info.getElasticsearchVersion().equals(currentElasticsearchVersion.toString()) == false) {
                 throw new IllegalArgumentException(
-                    "Stable Plugin ["
+                    "Plugin ["
                         + info.getName()
                         + "] was built for Elasticsearch version "
                         + info.getElasticsearchVersion()
-                        + " but earlier version "
-                        + Version.CURRENT
+                        + " but version "
+                        + Build.current().version()
                         + " is running"
                 );
             }
-        } else if (info.getElasticsearchVersion().equals(Version.CURRENT) == false) {
-            throw new IllegalArgumentException(
-                "Plugin ["
-                    + info.getName()
-                    + "] was built for Elasticsearch version "
-                    + info.getElasticsearchVersion()
-                    + " but version "
-                    + Version.CURRENT
-                    + " is running"
-            );
         }
         JarHell.checkJavaVersion(info.getName(), info.getJavaVersion());
     }
 
+    private record SemanticVersion(int major, int minor, int bugfix, String suffix) {
+
+        static final Pattern semanticPattern = Pattern.compile("^(\\d+)\\.(\\d+)\\.(\\d+)(\\D?.*)$");
+        static SemanticVersion create(String version) {
+            Matcher matcher = semanticPattern.matcher(version);
+            if (matcher.matches() == false) {
+                throw new IllegalArgumentException("Cannot be parsed as a semantic version: " + version);
```

Comment:
The control flow seems a bit odd, this exception is generated, but then it is thrown away. Can we move this method directly into the plugin utils and parse+validate at the same time, thereby allowing the exception message to be tailored to the context?

Reply:
Good point. I took a pass at this. Let me know if it's what you're looking for.